File: Backend/DTOs/ComponentSpecificationDTO.cs
using System.Text.Json.Serialization;

namespace CircuitBench.DTOs
{
    [JsonPolymorphic(TypeDiscriminatorPropertyName = "kind")]
    [JsonDerivedType(typeof(ResistorComponentSpecificationDTO), "resistor")]
    [JsonDerivedType(typeof(LEDComponentSpecificationDTO), "led")]
    public abstract record ComponentSpecificationDTO;

    public record ResistorComponentSpecificationDTO(double Resistance) : ComponentSpecificationDTO;

    public record LEDComponentSpecificationDTO(double ForwardVoltage, double MaxCurrent) : ComponentSpecificationDTO;
}

File: Frontend/src/dtos/component-specification-dto.ts
export interface LEDComponentSpecificationDTO {
  readonly kind: "led"
  readonly forwardVoltage: number
  readonly maxCurrent: number
}

export interface ResistorComponentSpecificationDTO {
  readonly kind: "resistor"
  readonly resistance: number
}

export type ComponentSpecificationDTO =
  LEDComponentSpecificationDTO | ResistorComponentSpecificationDTO
